### ISSUE TITLE
[RFC] Eager Mode Support on Top of torch.compile

### ISSUE THREAD

**Issue by EikanWang** [pytorch/pytorch#115545]
## Motivation

By now, PyTorch has defined > 2000 operations. Meanwhile, PyTorch users start from eager mode. If a new backend intends to support PyTorch eager mode, it means that the backend, like Intel GPU, has to implement all these operations. Otherwise, the users may encounter `unimplemented` errors if they run PyTorch on the new backend. There are two practical frictions here - engineering effort and maintenance effort. 

- Engineering Effort - The new backend has to implement all ATen operations(#of ATen operations >2000). The development efforts are dramatic.
- Maintenance Effort - PyTorch operations may change its interfaces, such as adding or removing parameters. Regarding these changes, the backend maintainers will need to adopt these backend-specific implementations according to these changes.

Given these challenges, we propose an alternate technical pathway for eager mode support through torch.compile, offering two main advantages over the traditional implementation approach:

- In terms of a particular backend, it needs to implement the fallback operations that `torch.compile` has not supported yet or `torch.compile` falls back to ATen explicitly.
-  `torch.compile` interface is device agnostic. A particular ATen operation does not need to be implemented for different backends but just by leveraging `torch.compile` to generate performant backend code.

## Approach

We propose compiling single aten operations using `torch.compile` and registering them in PyTorch on the Python side. This approach simplifies backend implementation and maintenance. Here's an illustrative example:

```python
# Example code demonstrating the proposed method
x = torch.empty(1).fill_(1)
y = torch.empty(1).fill_(2)

def wrapper_fn_sub(*args, **kwargs):
    with torch._C._SetExcludeDispatchKeyGuard(torch._C.DispatchKey.Python, False):
        opt_fn_sub = torch.compile(torch.ops.aten.sub)
        res = opt_fn_sub(a, b)
        return res

custom_op_lib_xpu_impl = torch.library.Library("aten", "IMPL")
custom_op_lib_xpu_impl.impl("sub.Tensor", wrapper_fn_sub, "CPU")
res = x - y

ref = torch.empty(1).fill_(-1)
assert all(res == ref)
```

## Current Status and Challenges

We are working on the exploration and have enabled the above example by providing another alternative registration API for POC. We will support more operations and both inference and training to check if there are more feature gaps.

Besides the feature implementation, there are some challenges.
- The compilation overhead may impact the user experience as the aten operations need to be compiled for the first time. 
- The feature requires robust dynamic support. Otherwise, the single aten operation has to be recompiled as long as the shapes of the input tensors are changed.

Regarding these challenges, we may address them by improving the persistent disk cache of `torch.compile` and dynamic support.

### Additional context

Currently, the Python kernel for a particular ATen operation will enable hermetic Python objects assumption. 
https://github.com/pytorch/pytorch/blob/b88be1686d1eaf52cd71ef19c81957312c3f4558/torch/csrc/utils/python_dispatch.cpp#L174

If a Python kernel invokes `torch.compile`, it will break the assumption. Take the above code snippet as an example, `torch.compile` will build `FakeTensor` but aten dispatch has enabled hermetic python object assumption in the aten dispatch path.

> While HermeticPyObject was enabled, we attempted to create a tensor subclass with __torch_dispatch__.  This violates the invariant that operations in HermeticPyObject have equivalent C++ implementations. If your operator registered from Python operator registration isn't doing anything strange, there may be an internal PyTorch bug involving not appropriately disabling TorchDispatchMode before executing Python op registration. https://github.com/pytorch/pytorch/blob/b88be1686d1eaf52cd71ef19c81957312c3f4558/torch/csrc/autograd/python_variable.cpp#L1962-L1970

To address the issue, we may provide a dedicated registration API to indicate a Python kernel does not enable hermetic python object assumption.

**Comment by EikanWang** [comment]
@malfet , do you mean the sparse operations have enabled such a feature? I will double-check it.

**Comment by EikanWang** [comment]
Currently, I can run some elementwise cases by providing a dedicated registration API like `impl_compile` besides other registration APIs to bypass `EnableHermeticPyObject` check. @jansel , any comments on the solution?

**Comment by ezyang** [comment]
cc @jbschlosser on hermetic. I'll pipe up here later

**Comment by zou3519** [comment]
It seems like there are two things going on here:
1. People want to use torch.compile to add new functions to PyTorch. That is, we would torch.compile some sequence of PyTorch operations, and maybe also add a fast cache on top of it, and expose this as a `torch.*` API.
2. People want to write aten operators whose implementations use torch.compile.

The main difference between (1) and (2) is if the final function is an aten operator (2) or something else (1).

It sounds like this issue is pursuing (2), but I want to challenge that for a bit. Why can't we just do (1)? Concretely, the proposal would be to have something like:
```py
@better_compile(fullgraph=True, backend="inductor")
def my_activation(x):
    return torch.where(x > 0, x ** 2, -0.5 * x)

# accessible via torch
torch.my_activation(x)
```
where `better_compile` is torch.compile, but perhaps with more limitations and a faster cache. This function (my_activation) would automatically support all PyTorch subsystems via AOTDispatcher (we would need to build this out) and as a plus, would not increase the number of aten operators in PyTorch (unlike option 2).

**Comment by jgong5** [comment]
> The main difference between (1) and (2) is if the final function is an aten operator (2) or something else (1).
> 
> It sounds like this issue is pursuing (2), but I want to challenge that for a bit. Why can't we just do (1)?

I see (1) is a superset of (2) where aten op is a special case of an arbitrary op of (1)? In general, I think (1) is a reasonable use case where one wants to implement a custom op efficiently by simply constructing a sequence of aten ops in a device-agnostic manner. As an example, we are proposing such a light-weight optimization to bitsandbytes for some quant and dequant ops: https://github.com/TimDettmers/bitsandbytes/issues/894. This would relieve us from manually implementing the native kernels. But the problems mentioned in the RFC still hold, in particular, the python and compilation overhead.

> This function (my_activation) would automatically support all PyTorch subsystems via AOTDispatcher (we would need to build this out) and as a plus, would not increase the number of aten operators in PyTorch (unlike option 2).

Why do you think option 2 would **increase** the number of aten operators? We do not intend to increase the number but just wanted to support existing aten operators.

**Comment by EikanWang** [comment]
In terms of (1), the custom function can be accessible via torch, but I'm kind of concerned it might not be able to support all cases. Suppose a code block is as follows. 

```python
import torch
a = torch.randn(10, device="xpu")
b = torch.randn(10, device="xpu")

// (EIKAN: I suppose the "add" here is not routed to the torch.)
c = a.add(b)
```

May I know the idea to support "add" w/o any code change?

**Comment by zou3519** [comment]
Okay, I think I see what's going on. Is this RFC for using torch.compile to implement new operators in PyTorch, or is it to help alternative backends (like XPU) gain coverage for existing PyTorch operations? 

Based on the replies, it sounds like torch.compile supports XPU, but we do not support XPU in eager-mode, so you want to somehow use torch.compile to add eager-mode support for XPU. Is my understanding correct?

**Comment by EikanWang** [comment]
> is it to help alternative backends (like XPU) gain coverage for existing PyTorch operations?

Yes

> Based on the replies, it sounds like torch.compile supports XPU, but we do not support XPU in eager-mode, so you want to somehow use torch.compile to add eager-mode support for XPU. Is my understanding correct?

Yes

> If this is the case, why are we registering the implementation to DispatchKey::CPU in the example? I would expect that we only register the torch.compile'd function to a DispatchKey for XPU (if that exists)

I am sorry to confuse you by taking the CPU dispatch key as an example, as XPU is still WIP and cannot verify the idea. I will update the example code to avoid confusion.

**Comment by EikanWang** [comment]
@zou3519 , thank you for pointing out the part that led to ambiguity in the idea. I have updated the example accordingly.

**Comment by jansel** [comment]
This was some related discussion on #116368

**Comment by ezyang** [comment]
I think https://github.com/pytorch/pytorch/pull/116996 will resolve the immediate tactical problem regarding hermetic mode

However, I still have some questions about the proposed implementation strategy here. In particular, how much overhead per eager mode call is acceptable overhead? A big reason why most of PyTorch moved into C++ was to reduce eager mode overhead. With the prototype flow here, we have to dispatch back into Python for each op, and that's going to be expensive. Do you have a target latency here?

**Comment by EikanWang** [comment]
> how much overhead per eager mode call is acceptable overhead?

I collected the overhead info (CPU is Intel(R) Xeon(R) Platinum 8260 CPU @ 2.40GHz; PyTorch commit: e8a9d088c694a49c414152ef6ea28b475e8f85f2)

- Initialization Phase:
There is a significant overhead of over **1 second** during the initial run of `torch.compile`. This overhead is a one-time occurrence, triggered only during the first compilation.

- Performance of Specific Aten Operations Post-Initialization:

  - On First Execution: After the initial `torch.compile` setup, the overhead for executing a specific Aten operation for the first time is approximately **115 milliseconds**.

  - On Subsequent Executions Without Re-compilation: When the same Aten operation is executed again, without re-compiling, the overhead becomes negligible, aligning with the performance of the traditional eager mode.

we plan to add a cache mechanism to mitigate the overhead.

**Comment by ezyang** [comment]
> the overhead becomes negligible 

Did you actually measure this? Because the overhead is definitely not negligible lol

**Comment by EikanWang** [comment]
> > the overhead becomes negligible
> 
> Did you actually measure this? Because the overhead is definitely not negligible lol

I used the Python profiler to collect the performance data. The subsequent execution w/o re-compilation was **1ms** for a simple case.  I used the `time` to get the execution time of its equivalent C++ ATen operation run. It was also about **1ms**. 

```python
# Invoke torch.compile 
x - y

sortby = SortKey.CUMULATIVE
with cProfile.Profile() as pr:

  # Invoke torch.compile 
  x - y
  
  s = io.StringIO()
  ps = pstats.Stats(pr, stream=s).sort_stats(sortby)
  ps.print_stats()
  print(s.getvalue())
```

Python profile data:
![image](https://github.com/pytorch/pytorch/assets/55483091/f2108d7d-1cbb-4c17-839c-03be3c63bf7c)

I can improve the profile accuracy to microseconds and then take another look.

**Comment by EikanWang** [comment]
By the way, there should be no big difference compared to the C++ ATen operation run if we add a cache mechanism. 

Cache hit: Python(user script) -> C++ kernel(The kernel encapsulates a python function w/ torch.compile) -> C++(AOTI)

**Comment by ezyang** [comment]
I don't expect pstats to have the resolution you need. Just run the thing in a loop N iterations and measure the overall time. And compare it with CPU tensor

**Comment by EikanWang** [comment]
Sorry for the late reply. I used the high resolution to measure the overhead on A10. And yes, the overhead is **NOT** negligible. @ezyang, thanks for correcting me. We are narrowing down the detailed overhead to get deep insights and will update here.

**Comment by ezyang** [comment]
A flow that would get good perf is if you register C++ functions to dispatcher, which directly call into CPP wrapper generated kernels. However, there is a missing piece which is guard dispatch. @anijain2305 C++ guards could be part of the solution here, but there's still some more glue that would need to be implemented since they aren't intended to be used this way.

**Comment by sujoysaraswati** [comment]
@EikanWang , there was a discussion on this on the thread https://dev-discuss.pytorch.org/t/set-of-ops-for-a-backend-to-register/1024/8 sometime back and Sherlock was suggesting an eager decomposition at that point. The torch.compile based approach suggested here is essentially achieving this decomposition right? Does it also mean that the ops that the backend will receive from eager execution with this will all be functional as well?

**Comment by EikanWang** [comment]
I quickly verified the C++ path by hardcoding some logic to proxy the C++ solution and collected the performance of a simple case(`sub`) w/ 1M float elements by running 100000 times on CUDA A10. Compared to the Python version, which directly calls into `torch.compile`, the performance would be significantly improved if the aten operation was dispatched to the C++ function, which directly calls into CPP-Wrapper-generated kernels. But it is still not as performant as ATen native implementation - 10%~30% slower. One of the performance gaps comes from the kernel efficiency. I will implement the POC and collect more performance data for more insights.

**Comment by EikanWang** [comment]
> The torch.compile based approach suggested here is essentially achieving this decomposition right? 

The relation between the `decomposition` and this approach should be weak. `decomposition` is kind of like a part of the `torch.compile` implementation.

> Does it also mean that the ops that the backend will receive from eager execution with this will all be functional as well?

Yes

**Comment by abrasumente233** [comment]
Hi @EikanWang,

I noticed the last activity on this RFC was from May 10, 2024. Is this still being worked on?

If not, did Intel shift focus to supporting eager mode via SYCL or another approach?

Would appreciate any updates on the current status and direction.

Thanks!

**Comment by EikanWang** [comment]
This RFC does not serve XPU(Intel GPU). It is a general solution to support ATen operation through torch.compile. We have landed some PRs to establish the infrastructure. 

For Intel GPU, we support eager mode via SYCL.
